go/ct/spc/specification.go

```diff
@@ -1839,6 +1956,75 @@ func logOp(n int) []Rule {
 	return rules
 }
 
+func selfDestructEffect(s *st.State, destinationColdCost, dynamicGas, refoundGas vm.Gas) {
+	// Behavior pre cancun: the current account is registered to be destroyed, and will be at the end of the current
+	// transaction. The transfer of the current balance to the given account cannot fail. In particular,
+	// the destination account code (if any) is not executed, or, if the account does not exist, the
+	// balance is still added to the given address.
+
+	// account to send the current balance to
+	destinationAccount := s.Stack.Pop().Bytes20be()
+	currentAccount := s.CallContext.AccountAddress
+	CurrentBalance := s.Accounts.GetBalance(currentAccount)
+
+	// If a positive balance is sent to an empty account, the dynamic gas is 25000. Otherwise there is
+	// no additional cost. An account is empty if its balance is 0, its nonce is 0 and it has no code.
+	dynamicCost := vm.Gas(0)
+	if !CurrentBalance.IsZero() && !s.Accounts.Exist(destinationAccount) {
+		dynamicCost += dynamicGas
+	}
+
+	dynamicCost += destinationColdCost
+
+	if s.Gas < dynamicCost {
+		s.Status = st.Failed
+		return
+	}
+	s.Gas -= dynamicCost
+	s.Status = st.Stopped
+
+	// PC should not increase, but rulesFor does it for all regular cases, so we counter it here.
+	s.Pc--
+
+	s.ReturnData = s.LastCallReturnData
+	s.GasRefund += refoundGas
+}
+
+func selfDestructEffectWithWarming(s *st.State, destinationColdCost, dynamicGas, refoundGas vm.Gas) {
+	// Behavior pre cancun: the current account is registered to be destroyed, and will be at the end of the current
+	// transaction. The transfer of the current balance to the given account cannot fail. In particular,
+	// the destination account code (if any) is not executed, or, if the account does not exist, the
+	// balance is still added to the given address.
+
+	// account to send the current balance to
+	destinationAccount := s.Stack.Pop().Bytes20be()
+	currentAccount := s.CallContext.AccountAddress
+	CurrentBalance := s.Accounts.GetBalance(currentAccount)
+
+	// If a positive balance is sent to an empty account, the dynamic gas is 25000. Otherwise there is
+	// no additional cost. An account is empty if its balance is 0, its nonce is 0 and it has no code.
+	dynamicCost := vm.Gas(0)
+	if !CurrentBalance.IsZero() && !s.Accounts.Exist(destinationAccount) {
+		dynamicCost += dynamicGas
+	}
+
+	dynamicCost += destinationColdCost
+
+	if s.Gas < dynamicCost {
+		s.Status = st.Failed
+		return
+	}
+	s.Gas -= dynamicCost
+	s.Accounts.MarkWarm(destinationAccount)
```

Comment:
This function is almost identical to the function above, except that the account is warmed up here.

I would merge them, and check in the function whether the revision is high enough to cause the warm-up.